### PR TITLE
Add keyboard interactions and screenreader support to Activated Menus & Activators

### PR DESCRIPTION
**Pull request by chrispymm**
This PR addresses tickets [#1361](https://trello.com/c/qYTQed5J/1631-activated-menu-keyboard-support) and [#2444](https://trello.com/c/SfDxPUZZ/2444-activated-menu-activator-keyboard-support)

This PR is pretty large (probably too large) and there's a fair amount going on:

1. Adds Keyboard Support
2. Removes jQuery UI menu
3. Re-organizes code into single-file-per-class methodology and into directories
4. Refactors the code to be more modern and enforce encapsulation

### Keyboard Support
**Activator**
* When focused on `menubutton` `<Enter>`, `<Space>` or `<DownArrow>` opens menu and places focus on first element.
* When focused on `menubutton` `<Up>` opens menu and places focus on last element.

**Menu**
* When focused on a `menuitem` `<DownArrow>` moves focus to next item, wrapping to first.
* When focused on a `menuitem` `<UpArrow>` moves focus to previous item, wrapping to last.
* If item has a submenu (`<ul role="menu">`) `<Enter>`, `<Space>` or `<RightArrow>` opens the submenu and moves focus to first item.
* When focus is within a submenu `<LeftArrow>` closes the submenu and returns focus to the parent `menuitem`
* When focus is within a `menu` `<Home>` moves focus to first item in menu
* When focus is within a `menu` `<End>` moves focus to last item in menu
* When focus is within a `menu` `<Esc>` closes the whole menu and returns focus to the activator.
* When focus is within a `menu` `<Tab>` closes the menu and moves focus to the next tabbable element in the `document`
* When focus is within a `menu` `<Shift> + <Tab>` closes the menu and moves focus to the previous tabbable element in the `document`

### Removing jQuery UI
Theoretically jQuery UI supports keyboard interactions, but it's implementation seemed lacking - it handled keyboard updates by manipulating the `activedescendant` attribute on the menu, which should be valid but seemed to result in no screenreader output at all.  

I also ran into issues with handling the combination of keyboard events and mouse interaction within submenus, and it was impossible to hold focus state and return it to the correct elements.  

In the end I was reimplementing so much of the jQuery menu and was battling it so much that it seemed easier to remove it. 

We still utilise the jQuery UI `position` widget to handle the positioning of the menus.

### Reorganization
Each class now has it's own file, and there are a couple of new classes (`ActivatedMenuItem` and `ActivatedMenuSubmenu`) to handle the keyboard interactions in a more granular fashion rather than doing it all in the parent `ActivatedMenu` class.

The classes have been moved into a new directory `/components/menus` in order to start bringing some structure to the JS files in the project.

### Refactoring
* The various menu classes have been refactored to have all code encapsulated within the class.  All class method calls (`ActiveMenu.methodName.call(this)`) have been replaced by private class methods.
* Pseudo-private properties (variables starting with `_`) have been reafctored to private class fields, and where appropriate getter methods have been added.
* Across the classes methods have been renamed or introduced to have some consistency across all of the menu files.

N.B. The ActivatedMenu subclasses have not been as extensively refactored, as this had already turned into far too large a PR! (Sorry).

### CODE REVIEW

File: acceptance/features/edit_single_radios_question_page_spec.rb [hunk 1]
```diff
@@ -51,7 +51,7 @@
     given_I_have_a_single_question_page_with_radio
     when_I_update_the_question_name
     and_I_update_the_options
-    and_I_want_to_delete_an_option('PDFinn')
+    and_I_want_to_delete_an_option('Adobe-wan Kenobi')
```
Comment (brenetic) [inline acceptance/features/edit_single_radios_question_page_spec.rb:54]: kenobi > finn

File: acceptance/support/capybara.rb [hunk 2]
```diff
@@ -2,8 +2,9 @@
   chrome_options = Selenium::WebDriver::Chrome::Options.new.tap do |o|
     o.add_argument '--headless'
     o.add_argument '--no-sandbox'
+    o.add_argument '--window-size=1920,1200'
```
Comment (brenetic) [inline acceptance/support/capybara.rb:5]: 👍

File: acceptance/support/capybara.rb [hunk 3]
```diff
@@ -2,8 +2,9 @@
...
   end
-  Capybara::Selenium::Driver.new(app, browser: :chrome, options: chrome_options)
+  Capybara::Selenium::Driver.new(app, browser: :chrome, capabilities: chrome_options)
```
Comment (brenetic) [inline acceptance/support/capybara.rb:7]: awesome, thanks for doing that

File: app/javascript/src/components/menus/editable_collection_item_menu.js [hunk 4]
```diff
@@ -35,13 +34,14 @@ class EditableCollectionItemMenu extends ActivatedMenu {
       this.selection(event, ui.item);
     });
 
-    let $target = this._config.$target;
+    let $target = this.config.$target;
```
Comment (stevenburnell-moj) [inline app/javascript/src/components/menus/editable_collection_item_menu.js:37]: Although the original alluded to config being private while accepting JS limitation of keeping it public, with the underscore name prefix now removed we've also removed that indication it should be private. Is that intentional? The inherited ActivatedMenu class has done the full job of making it actually private (with modern hash naming), so I'm guessing this should be similar.

Again, not going to hold approval of PR but might be worth making a note to revisit when the dust settles.

**Review by stevenburnell-moj — APPROVED** [review]

File: app/javascript/src/component_dialog_form.js [hunk 5]
```diff
@@ -79,7 +79,17 @@ class FormDialog {
...
+    var $node = this.$node;
     this.$node.dialog("open");
+    window.setTimeout(function() {
+      // Not great but works.
+      // We want the focus put inside dialog as all functionality to trap tabbing is there already.
+      // Because we sometimes open dialogs from other components, those other components may (like
+      // menus) shift focus from the opening dialog. We need this delay to allow those other events
+      // to play out before we try to set focus in the dialog. Delay time is arbitrary but we
+      // obviously want it as low as possible to avoid user annoyance. Increase only if have to.
+      $node.parent().find("input:not([type='hidden']), button").not(".ui-dialog-titlebar-close").eq(0).focus();
```
[the remark this answers is not in the document]
Comment (chrispymm) [reply app/javascript/src/component_dialog_form.js:91]: We do I think IIRC all the others have it... it was just this one that didn't - hence the addition.

File: app/javascript/src/components/menus/editable_collection_item_menu.js (hunk 4 above)
Comment (chrispymm) [reply app/javascript/src/components/menus/editable_collection_item_menu.js:37]: Ah no, I looked further.  This is in the constructor so its using the passed in config directly.  This menu never actually assigns the passed in `config` to an internal `#config` var as it's not needed.

**Pull request merged**